### PR TITLE
Add StatsD formatter and reporter

### PR DESCRIPTION
**Pull request by Arkatufus**
### The issue or feature being addressed

#624

### Details on the issue fix or feature implementation

- Implement StatsD string formatter
- Use the socket reporting plug-in to implement TCP and UDP StatsD client
- Implement unit testing
- Implement console and web mvc sandbox projects

### Confirm the following

- [x] I have ensured that I have merged the latest changes from the dev branch
- [ ] I have successfully run a [local build](https://github.com/AppMetrics/AppMetrics#how-to-build)
- [x] I have included unit tests for the issue/feature
- [x] I have included the github issue number in my commits

### CODE REVIEW

File: src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDFormatterConstants.cs
Comment (alhardy) [inline src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDFormatterConstants.cs:15]: Small thing, class name from copy/paste?

File: src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs
Comment (alhardy) [inline src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs:94]: @Arkatufus not sure I'm following the sampling concept, could you please provide a bit more context?

**Review by alhardy — COMMENTED** [review]
LGTM, would just like a bit more concept around the sampling with StatsD

File: src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDFormatterConstants.cs
Comment (Arkatufus) [reply src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDFormatterConstants.cs:15]: Oh, right, i'll fix that in a jiffy.

File: src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs
Comment (Arkatufus) [reply src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs:94]: In StatsD, counter, histogram, and timer metrics can be undersampled by what they call sample rate and it is expressed in percent value. For example, a sample rate of 0.1 means that only 10% of the metric points emitted by the system will be captured and sent to the server; the server then extrapolate the approximate number by scaling the reported numbers.

In this implementation. sample rate can be set as default for all applicable metric type in the `MetricsStatsDOptions` option class or it can be set per metric measurement by adding a "sampleRate" tag to the measurement itself.

In this implementation, the sample filter `StatsDPointSampler` doubles as a StatsD metric name generator. 
1. Generate a unique name for each point. I'm partially following the Datadog implementation for this. The sampled point name is formatted as `[context].[measurement name, spaces replaced with underscore].[app metric metric type].[app metric item key value pair].[field name]`. This name is also used to send the sampled point to the StatsD server.
2. It then checks to see if undersampling can be used for the metric type, if not then it will simply stores it in a write queue  and returns.
3. The generated point name is used as a key to a dictionary of `<string, StatsDSampler>` to store/retrieve a simple sampler instance. I'll explain how the sampler work below.
4. Depending on the point sampler boolean return value, the point is either discarded or stored inside the write queue.
5. All of the points stored inside the queue are then written to the stream when `WriteAsync()` is called.

The point sampler `StatsDSampler` is quite simple. It uses an accumulator with type `double` and adds the `SampleRate` percent value to the accumulator every time the `Sample()` function is called. If the accumulator is greater or equal to one, subtract one from the accumulator and return true, else returns false. The sample point timestamp is used to differentiate each samples with each other, all samples that are captured in the same timestamp are considered as a single measurement and sent to the server.

Comment (alhardy) [reply src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs]: Thanks for the PR and taking the time to provide that context.

DD Ref: https://docs.datadoghq.com/developers/metrics/dogstatsd_metrics_submission/#sample-rates

Comment (Arkatufus) [reply src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs]: To be honest, it would be great to have a second pair of eyes to check if this implementation is correct or not. I have a nagging suspicion that I have missed something.

File: src/Reporting/src/App.Metrics.Formatters.Prometheus/Internal/Extensions/MetricSetItemExtensions.cs
Comment (Aaronontheweb) [inline src/Reporting/src/App.Metrics.Formatters.Prometheus/Internal/Extensions/MetricSetItemExtensions.cs:39]: Nice - like the addition of the parent tags

File: src/Reporting/src/App.Metrics.Formatting.StatsD/Builder/MetricsStatsDFormatterBuilder.cs
Comment (Aaronontheweb) [inline src/Reporting/src/App.Metrics.Formatting.StatsD/Builder/MetricsStatsDFormatterBuilder.cs:54]: Nitpick: StatsD is an OSS standard, but DataDog supports it as an interface.

File: src/Reporting/src/App.Metrics.Formatting.StatsD/DefaultStatsDMetricStringSerializer.cs
Comment (Aaronontheweb) [inline src/Reporting/src/App.Metrics.Formatting.StatsD/DefaultStatsDMetricStringSerializer.cs:22]: looks correct so far

File: src/Reporting/src/App.Metrics.Formatting.StatsD/DefaultStatsDMetricStringSerializer.cs
Comment (Aaronontheweb) [inline src/Reporting/src/App.Metrics.Formatting.StatsD/DefaultStatsDMetricStringSerializer.cs:27]: sampling serialization looks correct

File: src/Reporting/src/App.Metrics.Formatting.StatsD/DefaultStatsDMetricStringSerializer.cs
Comment (Aaronontheweb) [inline src/Reporting/src/App.Metrics.Formatting.StatsD/DefaultStatsDMetricStringSerializer.cs:49]: serializer looks good to me

File: src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs
Comment (Aaronontheweb) [reply src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs]: Can you write a simple spec here for each metric type - set sample rate at 10%, observe the number of attempted data points added (i.e. a reasonably large value) and then test to see that the value reported to the server falls within an acceptable range (using some kind of tolerance / margin) ? That will be more effective than reviewing the code here to find an error case. I think the code looks fine, but a test will be determinate.

**Review by Aaronontheweb — APPROVED** [review]
LGTM - but we should add a spec to make sure the sampling output is correct prior to being transmitted to the server.

File: src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs
Comment (alhardy) [reply src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs]: Question, StatsD sampling is there to reduce network traffic and load on the server i.e. it wouldn't be ideal doing a UDP call on every metric write for client or server.

AppMetrics already aggregates/samples on metric write in memory (on a side note, has some drawbacks), the StatsD reporter implementation here runs on a timer say every x secs. Is it still necessary to apply sampling on top of that?

Comment (alhardy) [reply src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs]: Sort of off topic to this PR, but, it would also be interesting for me, and most probably useful for others, to understand the drivers behind you guys looking at StatsD e.g. why not prom format as most monitoring agents support scraping prom format. Is it just that you want a push based abstraction?

File: src/Reporting/src/App.Metrics.Formatting.StatsD/Builder/MetricsStatsDFormatterBuilder.cs
Comment (Arkatufus) [reply src/Reporting/src/App.Metrics.Formatting.StatsD/Builder/MetricsStatsDFormatterBuilder.cs:54]: I'm having trouble finding the exact DogStatsD wire format. I think DogStatsD added tagging and events, I've added the tag support to this implementation (by appending `|#[tag1]:[value1],...[tagN]:[valueN]` at the end), but I haven't checked their implementation of event sending through StatsD.

Maybe we should differentiate the writer into more specialized writers instead of implementing a generic format writer? The default should work with vanilla StatsD, and have a dedicated writer for DogStatsD, and maybe add other formats later?

**Comment by Arkatufus** [comment]
Will do, I'll try to find a way to write one down.

File: src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs
Comment (Arkatufus) [reply src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs]: > AppMetrics already aggregates/samples on metric write in memory (on a side note, has some drawbacks), the StatsD reporter implementation here runs on a timer say every x secs. Is it still necessary to apply sampling on top of that?

Not really, but I'm trying to implement it as spec. The problem with vanilla StatsD is that its taking "real-time metrics collection" quite literally, ie. you can't really feed StatsD with a timestamped data, unless you make a custom StatsD server that understands how to extract and use the timestamp data to order the incoming data batch, or else it just treats the batch as a single lump of metric data at the time the data is received.

So in order to use StatsD with AppMetrics effectively, the data has to basically try to skip all internal batching and be sent to the StatsD server as soon as it generated as possible.

Comment (Arkatufus) [reply src/Reporting/src/App.Metrics.Formatting.StatsD/Internal/StatsDPointSampler.cs]: > why not prom format as most monitoring agents support scraping prom format. Is it just that you want a push based abstraction?

Well, we're basing one of our reporting products on AppMetrics and a few of our customers asked if we can implement StatsD as one of the transport protocol.

File: src/Reporting/src/App.Metrics.Formatting.StatsD/Builder/MetricsStatsDFormatterBuilder.cs
Comment (Arkatufus) [reply src/Reporting/src/App.Metrics.Formatting.StatsD/Builder/MetricsStatsDFormatterBuilder.cs:54]: Note that this tagging format is for AWS StatsD format, which might be the same as the one DogStatsD is using.

**Comment by Aaronontheweb** [comment]
What are we waiting on before this can be merged?

**Comment by Arkatufus** [comment]
- Revamped how sampling is done, following the official StatsD implementation
- Simplify Options class
- Split serializer into a vanilla StatsD implementation and DataDog DogStatsD implementation
- Optimize wire format by pruning a lot of non-needed data from the sent UDP packet.

**Comment by Arkatufus** [comment]
Unit test will probably fail because of App.Metrics.Abstraction.dll version bug.

**Comment by Aaronontheweb** [comment]
Looks like everything passed - how do you feel about it @alhardy ?

**Comment by alhardy** [comment]
> Looks like everything passed - how do you feel about it @alhardy ?

I think it is a little odd using StatsD with a library like AppMetrics which is already batching, but I guess it is just another way of transporting metrics at the end of the day. 

Implementation LGTM, appreciate the contribution 👍

**Pull request merged**
